src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,152 +1,119 @@
-    {
-        private readonly ImmutableArray<IOperation>.Builder _statements;
-        private readonly ImmutableHashSet<BasicBlock>.Builder _predecessors;
-        internal (IOperation Value, Branch Branch) InternalNext;
-        internal (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
-
-        public BasicBlock(BasicBlockKind kind)
-        {
-            Kind = kind;
-            _statements = ImmutableArray.CreateBuilder<IOperation>();
-            _predecessors = ImmutableHashSet.CreateBuilder<BasicBlock>();
-        }
-
-        public BasicBlockKind Kind { get; private set; }
-        public ImmutableArray<IOperation> Statements => _statements.ToImmutable();
-
-        /// <summary>
-        /// PROTOTYPE(dataflow): Tuple is temporary return type, we probably should use special structure instead.
-        /// </summary>
-        public (IOperation Condition, bool JumpIfTrue, Branch Branch) Conditional => InternalConditional;
-
         /// <summary>
-        /// PROTOTYPE(dataflow): During CR there was a suggestion to use different name - "Successor".
+        /// PROTOTYPE(dataflow): Rename this type to "BasicBlockBuilder" and source file to "ControlFlowGraphBuilder.BasicBlockBuilder.cs"
         /// </summary>
-        public (IOperation Value, Branch Branch) Next => InternalNext;
-
-        public ImmutableHashSet<BasicBlock> Predecessors => _predecessors.ToImmutable();
+        internal sealed class BasicBlock
+        {
+            private readonly ArrayBuilder<IOperation> _statementsBuilder;
+            private readonly HashSet<BasicBlock> _predecessorsBuilder;
+            public (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
+            public (IOperation Value, Branch Branch) InternalNext;
+            
+            public BasicBlock(BasicBlockKind kind)
+            {
+                Kind = kind;
+                _statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
+                _predecessorsBuilder = new HashSet<BasicBlock>();
+            }
 
-        public int Ordinal { get; internal set; } = -1;
+            public IReadOnlyCollection<IOperation> Statements => _statementsBuilder;
+            public ImmutableArray<IOperation> GetImmutableStatements() => _statementsBuilder.ToImmutableAndFree();
+            public ISet<BasicBlock> Predecessors => _predecessorsBuilder;
+            public BasicBlockKind Kind { get; }
 
-        public bool IsReachable { get; internal set; } = false;
+            public int Ordinal { get; set; } = -1;
 
-        /// <summary>
-        /// Enclosing region
-        /// </summary>
-        public ControlFlowGraph.Region Region { get; internal set; }
+            public bool IsReachable { get; set; } = false;
 
-        internal void AddStatement(IOperation statement)
-        {
-            _statements.Add(statement);
-        }
+            public ControlFlowRegion Region { get; set; }
```

Comment:
>public ControlFlowRegion Region [](start = 12, length = 31)

Why are some `public` members properties while others are fields? Could we use fields for all?

Reply:
@AlekseyTs is planning to clean up this type further (value and condition unification), so I will let him clean this up too.

---
In reply to: [191912797](https://github.com/dotnet/roslyn/pull/27254#discussion_r191912797) [](ancestors = 191912797)

Reply:
Yes, I will follow up with additional changes to this type

---
In reply to: [192183074](https://github.com/dotnet/roslyn/pull/27254#discussion_r192183074) [](ancestors = 192183074,191912797)